client/state/comments/reducer.js

```diff
@@ -24,169 +24,165 @@ import {
+export function items( state = {}, action ) {
+	const { siteId, postId, commentId, skipSort } = action;
+
 	switch ( action.type ) {
 		case COMMENTS_CHANGE_STATUS:
-			return updateSpecificState( state, action, ( comments = Immutable.List() ) => {
-				updateExistingIn(
-					comments,
-					comment => comment.get( 'ID' ) === action.commentId,
-					comment => comment.set( 'status', action.status )
-				);
-			} );
+			const { status } = action;
+			return {
+				...state,
+				[ `${ siteId }-${ postId }` ]: map( state[ `${ siteId }-${ postId }` ], comment => {
+					if ( comment.ID !== commentId ) {
+						return comment;
+					}
+
+					return {
+						...comment,
+						status
+					};
+				} )
+			};
 		case COMMENTS_EDIT:
-			return updateSpecificState( state, action, ( comments = Immutable.List() ) => {
-				updateExistingIn(
-					comments,
-					comment => comment.get( 'ID' ) === action.commentId,
-					comment => comment.set( 'content', action.content )
-				);
-			} );
+			const { content } = action;
+			return {
+				...state,
+				[ `${ siteId }-${ postId }` ]: map( state[ `${ siteId }-${ postId }` ], comment => {
+					if ( comment.ID !== commentId ) {
+						return comment;
+					}
+
+					return {
+						...comment,
+						content
+					};
+				} )
+			};
 		case COMMENTS_RECEIVE:
-			// create set of ids for faster lookup for filter later
-			const newIds = Immutable.Set( action.comments.map( comment => comment.ID ) );
-
-			// we prefer freshly retrieved data, so throw away old data
-			return updateSpecificState( state, action, function( comments = Immutable.List() ) {
-				let newComments = comments.filter( ( comment ) => ! newIds.has( comment.get( 'ID' ) ) )
-											.concat( Immutable.fromJS( action.comments ) );
-
-				if ( ! action.skipSort ) {
-					newComments = newComments.sort( ( a, b ) => new Date( b.get( 'date' ) ) - new Date( a.get( 'date' ) ) );
-				}
-
-				return newComments;
-			} );
-
+			return {
+				...state,
+				[ `${ siteId }-${ postId }` ]: ! skipSort ? sortBy( action.comments, [ 'date' ] ) : action.comments
+			};
 		case COMMENTS_REMOVE:
-			return updateSpecificState( state, action, ( comments = Immutable.List() ) =>
-				comments.filter( ( comment ) => comment.get( 'ID' ) !== action.commentId )
-			);
+			return {
+				...state,
+				[ `${ siteId }-${ postId }` ]: filter( state[ `${ siteId }-${ postId }` ], { ID: commentId } )
+			};
 		case COMMENTS_LIKE:
-			return updateSpecificState( state, action, ( comments = Immutable.List() ) =>
-				updateExistingIn( comments,
-					( comment ) => comment.get( 'ID' ) === action.commentId,
-					getLikeSetter( { like: true } )
-				)
-			);
+			return {
+				...state,
+				[ `${ siteId }-${ postId }` ]: map( state[ `${ siteId }-${ postId }` ], comment => {
+					if ( comment.ID !== commentId ) {
+						return comment;
+					}
+
+					return {
+						...comment,
+						i_like: true,
+						like_count: comment.like_count++
```

Comment:
If the tests aren't breaking from this, let's add `deepFreeze` to the test cases.